```
Nama  : Vicky Eldora Wuisan

Pembersihan data dari seperti cleaning data, merapikan data, normalisasi data, sehingga nantinya data siap dimasukkan ke dalam database, dan kemudian dibuatkan database dengan PostgreSQL server.
```

# 1. Data Exploration

## a. Load file data dengan Pandas lalu tampilkan data. Informasi apa yang bisa kamu sampaikan dari langkah ini?

- Dapat mengetahui jumlah kolom dan baris
- Dapat mengetahui nama kolom
- Dapat mengetahui values
- Dapat mengetahui type data

In [147]:
import pandas as pd

In [148]:
# mengimport data csv ke notebook
file = 'https://raw.githubusercontent.com/FTDS-learning-materials/phase-0/main/src/Financials.csv'
df = pd.read_csv(file)

In [149]:
# menampilkan tabel
df

Segment                   Country     Product   Discount Band   \
0          Government                    Canada   Carretera            None    
1          Government                   Germany   Carretera            None    
2           Midmarket                    France   Carretera            None    
3           Midmarket                   Germany   Carretera            None    
4           Midmarket                    Mexico   Carretera            None    
..                ...                       ...          ...             ...   
695    Small Business                    France    Amarilla            High    
696    Small Business                    Mexico    Amarilla            High    
697        Government                    Mexico     Montana            High    
698        Government                    Canada       Paseo            High    
699  Channel Partners  United States of America         VTT            High    

     Units Sold   Manufacturing Price   Sale Price     Gross Sales   \
0     $1,618.50                 $3.00       $20.00      $32,370.00    
1     $1,321.00                 $3.00       $20.00      $26,420.00    
2     $2,178.00                 $3.00       $15.00      $32,670.00    
3       $888.00                 $3.00       $15.00      $13,320.00    
4     $2,470.00                 $3.00       $15.00      $37,050.00    
..           ...                   ...          ...             ...   
695   $2,475.00               $260.00      $300.00    $7,42,500.00    
696     $546.00               $260.00      $300.00    $1,63,800.00    
697   $1,368.00                 $5.00        $7.00       $9,576.00    
698     $723.00                $10.00        $7.00       $5,061.00    
699   $1,806.00               $250.00       $12.00      $21,672.00    

         Discounts           Sales            COGS        Profit         Date  \
0             $-        $32,370.00      $16,185.00    $16,185.00   01/01/2014   
1             $-        $26,420.00      $13,210.00    $13,210.00   01/01/2014   
2             $-        $32,670.00      $21,780.00    $10,890.00   01/06/2014   
3             $-        $13,320.00       $8,880.00     $4,440.00   01/06/2014   
4             $-        $37,050.00      $24,700.00    $12,350.00   01/06/2014   
..              ...             ...             ...           ...         ...   
695   $1,11,375.00    $6,31,125.00    $6,18,750.00    $12,375.00   01/03/2014   
696     $24,570.00    $1,39,230.00    $1,36,500.00     $2,730.00   01/10/2014   
697      $1,436.40       $8,139.60       $6,840.00     $1,299.60   01/02/2014   
698        $759.15       $4,301.85       $3,615.00       $686.85   01/04/2014   
699      $3,250.80      $18,421.20       $5,418.00    $13,003.20   01/05/2014   

     Month Number  Month Name   Year  
0               1     January   2014  
1               1     January   2014  
2               6        June   2014  
3               6        June   2014  
4               6        June   2014  
..            ...          ...   ...  
695             3       March   2014  
696            10     October   2014  
697             2    February   2014  
698             4       April   2014  
699             5         May   2014  

[700 rows x 16 columns]

## b. Tampilkan informasi summary data dan berikan penjelasan mengenai temuan yang diperoleh.

- temuan yang diperoleh adalah ringkasan singkat tentang data yang dimiliki, seperti rangeindex, total kolom, nama kolom, berupa jumlah nilai bukan nol di setiap kolom dan tipe data kolom

In [150]:
# menampilkan info data tabel
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Segment                700 non-null    object
 1   Country                700 non-null    object
 2    Product               700 non-null    object
 3    Discount Band         700 non-null    object
 4    Units Sold            700 non-null    object
 5    Manufacturing Price   700 non-null    object
 6    Sale Price            700 non-null    object
 7    Gross Sales           700 non-null    object
 8    Discounts             700 non-null    object
 9     Sales                700 non-null    object
 10   COGS                  700 non-null    object
 11   Profit                700 non-null    object
 12  Date                   700 non-null    object
 13  Month Number           700 non-null    int64 
 14   Month Name            700 non-null    object
 15  Year                   

## c. Tampilkan list nama-nama kolom dan berikan penjelasan temuan apa yang kamu dapatkan, serta berikan langkah apa selanjutnya dari temuan ini.

- setelah melihat hasilnya ada nama kolom yang terdapat spasi di awal dan akhir sehingga perlu dilakukan langkah selanjutnya agar nama kolom menjadi lebih rapih yaitu harus memperbaiki penulisan kolom dengan menghilangkan spasi yang tidak diperlukan

In [151]:
# menampilkan nama kolom
df.columns

Index(['Segment', 'Country', ' Product ', ' Discount Band ', ' Units Sold ',
       ' Manufacturing Price ', ' Sale Price ', ' Gross Sales ', ' Discounts ',
       '  Sales ', ' COGS ', ' Profit ', 'Date', 'Month Number',
       ' Month Name ', 'Year'],
      dtype='object')

## d. Tampilkan list nilai unik masing-masing kolom `Segment`, `Country`, `Product`, `Discount Band` dan berikan informasi apa yang dapat kamu ambil dari langkah ini.

- dari langkah ini saya dapat mengetahui apa saja unique value dari kolom tersebut tanpa perlu harus melihat data tabel secara manual

In [152]:
# menampilkan value unik dari tabel segment
df['Segment'].unique()

array(['Government', 'Midmarket', 'Channel Partners', 'Enterprise',
       'Small Business'], dtype=object)

In [153]:
# menampilkan value unik dari tabel country
df['Country'].unique()

array(['Canada', 'Germany', 'France', 'Mexico',
       'United States of America'], dtype=object)

In [154]:
# menampilkan value unik dari tabel product
df[' Product '].unique()

array([' Carretera ', ' Montana ', ' Paseo ', ' Velo ', ' VTT ',
       ' Amarilla '], dtype=object)

In [155]:
# menampilkan value unik dari tabel discount
df[' Discount Band '].unique()

array([' None ', ' Low ', ' Medium ', ' High '], dtype=object)

# 2. Data Cleaning

## a. Manipulasi kolom meliputi penggantian nama kolom, penghapusan kolom, atau penambahan kolom

In [156]:
# menghilangkan spasi yang ada awal dan akhir nama kolom
df.columns= df.columns.str.strip()

In [157]:
# menampilkan semua nama kolom
df.columns

Index(['Segment', 'Country', 'Product', 'Discount Band', 'Units Sold',
       'Manufacturing Price', 'Sale Price', 'Gross Sales', 'Discounts',
       'Sales', 'COGS', 'Profit', 'Date', 'Month Number', 'Month Name',
       'Year'],
      dtype='object')

## b. Penggantian tipe data yang sesuai dengan seharusnya. Untuk nilai harga dengan simbol mata uang, simbol tersebut dapat dibuang supaya dapat terkonversi ke tipe data yang seharusnya. Perhatikan format penulisan angkanya apakah ada tanda yang membuat angka hanya bisa bertipe data string. (Tambahan: Di suatu kolom yang bernilai uang, terpadat beberapa data yang mengandung karakter '(' dan ')' di dalamnya sehingga harus ditake-out karakter tersebut).

## c. Cek dan handling missing value. (Hint: terdapat non-standard missing value pada kolom yang semestinya bertipe data integer, handling dapat dilakukan dengan imputasi dengan angka nol '0')

- format dari setiap kolom masih belum tepat sehingga belum bisa dilakukan data analysis. 
- kolom data `date` masih bertipe obeject, dimana akan lebih sesuai jika formatnya date time.
(dengan melakukan pd.to_datetime)
- kolom `Unit Sold`,`Manufacturing Price`, `Sale Price`,`Gross Sales`,`Discounts`,`Sales`,`COGS`,`Profit` masih bertipe object, dimana kurang tepat, dan lebih tepat jika formatnya int atau float.
(pertama dilakukan penghilangan karaktek '$', ',', ' ', '(', ')', '-'. kemudian mengubah data type dengan as type)

In [158]:
# melakukan replace data menjadi kosong agar
df['Units Sold'] = df['Units Sold'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','')
df['Manufacturing Price'] = df['Manufacturing Price'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','')
df['Sale Price'] = df['Sale Price'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','')
df['Gross Sales'] = df['Gross Sales'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','')
df['Discounts'] = df['Discounts'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','0')
df['Sales'] = df['Sales'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','')
df['COGS'] = df['COGS'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','')
df['Profit']= df['Profit'].str.replace('$','').str.replace(',','').str.replace('(','').str.replace(')','').str.replace(' ','').str.replace('-','0')

In [159]:
# menampilkan ringkasan singkat tentang data saya, rangeindex, total kolom, nama kolom, berupa jumlah nilai bukan nol di setiap kolom dan tipe data kolom
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Segment              700 non-null    object
 1   Country              700 non-null    object
 2   Product              700 non-null    object
 3   Discount Band        700 non-null    object
 4   Units Sold           700 non-null    object
 5   Manufacturing Price  700 non-null    object
 6   Sale Price           700 non-null    object
 7   Gross Sales          700 non-null    object
 8   Discounts            700 non-null    object
 9   Sales                700 non-null    object
 10  COGS                 700 non-null    object
 11  Profit               700 non-null    object
 12  Date                 700 non-null    object
 13  Month Number         700 non-null    int64 
 14  Month Name           700 non-null    object
 15  Year                 700 non-null    int64 
dtypes: int64

In [160]:
# merubah data type dari object menjadi float
df['Units Sold']  = df['Units Sold'].astype(float)
df['Manufacturing Price']  = df['Manufacturing Price'].astype(float)
df['Sale Price'] = df['Sale Price'].astype(float)
df['Gross Sales'] = df['Gross Sales'].astype(float)
df['Discounts'] = df['Discounts'].astype(float)
df['Sales'] = df['Sales'].astype(float)
df['COGS'] = df['COGS'].astype(float)
df['Profit']= df['Profit'].astype(float)

In [161]:
# merubah data type dari object menjadi date
df['Date'] = pd.to_datetime(df['Date'])
df['Month Number'] = pd.to_datetime(df['Month Number'])

# 3. Data Normalization

## a. Lakukan normalisasi data yang sudah diolah sebelumnya sampai dengan 3NF.

## b. Jumlah tabel setelah dinormalisasi sebanyak 5 tabel termasuk tabel utama.

## c. Perhatikan penggunaan primary key dan foreign key antar tabel.

## d. Kemudian simpan semua tabel yang dihasilkan ke dalam file csv. Berikan nama file sesuai dengan nama tabel yang akan digunakan pada database. Jangan lupa index dataframe tidak perlu diikutsertakan dalam file karena skema tabel akan menerapkan auto incremental untuk pengisian primay key

In [162]:
# Mengecek apakah ada value yang sama, kemudian hapus row value yang sama, tapi karena pada tabel ini tidak ada yang sama (sebenernya tidak usah melakukan ini)
df.duplicated()
df.drop_duplicates()

Segment                   Country      Product Discount Band  \
0          Government                    Canada   Carretera          None    
1          Government                   Germany   Carretera          None    
2           Midmarket                    France   Carretera          None    
3           Midmarket                   Germany   Carretera          None    
4           Midmarket                    Mexico   Carretera          None    
..                ...                       ...          ...           ...   
695    Small Business                    France    Amarilla          High    
696    Small Business                    Mexico    Amarilla          High    
697        Government                    Mexico     Montana          High    
698        Government                    Canada       Paseo          High    
699  Channel Partners  United States of America         VTT          High    

     Units Sold  Manufacturing Price  Sale Price  Gross Sales  Discounts  \
0        1618.5                  3.0        20.0      32370.0       0.00   
1        1321.0                  3.0        20.0      26420.0       0.00   
2        2178.0                  3.0        15.0      32670.0       0.00   
3         888.0                  3.0        15.0      13320.0       0.00   
4        2470.0                  3.0        15.0      37050.0       0.00   
..          ...                  ...         ...          ...        ...   
695      2475.0                260.0       300.0     742500.0  111375.00   
696       546.0                260.0       300.0     163800.0   24570.00   
697      1368.0                  5.0         7.0       9576.0    1436.40   
698       723.0                 10.0         7.0       5061.0     759.15   
699      1806.0                250.0        12.0      21672.0    3250.80   

         Sales      COGS    Profit       Date                  Month Number  \
0     32370.00   16185.0  16185.00 2014-01-01 1970-01-01 00:00:00.000000001   
1     26420.00   13210.0  13210.00 2014-01-01 1970-01-01 00:00:00.000000001   
2     32670.00   21780.0  10890.00 2014-01-06 1970-01-01 00:00:00.000000006   
3     13320.00    8880.0   4440.00 2014-01-06 1970-01-01 00:00:00.000000006   
4     37050.00   24700.0  12350.00 2014-01-06 1970-01-01 00:00:00.000000006   
..         ...       ...       ...        ...                           ...   
695  631125.00  618750.0  12375.00 2014-01-03 1970-01-01 00:00:00.000000003   
696  139230.00  136500.0   2730.00 2014-01-10 1970-01-01 00:00:00.000000010   
697    8139.60    6840.0   1299.60 2014-01-02 1970-01-01 00:00:00.000000002   
698    4301.85    3615.0    686.85 2014-01-04 1970-01-01 00:00:00.000000004   
699   18421.20    5418.0  13003.20 2014-01-05 1970-01-01 00:00:00.000000005   

     Month Name  Year  
0      January   2014  
1      January   2014  
2         June   2014  
3         June   2014  
4         June   2014  
..          ...   ...  
695      March   2014  
696    October   2014  
697   February   2014  
698      April   2014  
699        May   2014  

[700 rows x 16 columns]

In [163]:
# untuk membuat data frame kemudian dalam data frame tersebut hanya berisikan value unik dari kolom tersebut
Segment_table = pd.DataFrame(df['Segment'].unique())
Country_table = pd.DataFrame(df['Country'].unique())
Product_table = pd.DataFrame(df['Product'].unique()) 
Discount_Band_table = pd.DataFrame(df['Discount Band'].unique())

In [164]:
# menampilkan tabel
Segment_table

0
0        Government
1         Midmarket
2  Channel Partners
3        Enterprise
4    Small Business

In [165]:
# menampilkan tabel
Country_table

0
0                    Canada
1                   Germany
2                    France
3                    Mexico
4  United States of America

In [166]:
# menampilkan tabel
Product_table

0
0   Carretera 
1     Montana 
2       Paseo 
3        Velo 
4         VTT 
5    Amarilla

In [167]:
# menampilkan tabel
Discount_Band_table

0
0     None 
1      Low 
2   Medium 
3     High

In [168]:
# mengsave data diatas menjadi csv
Segment_table.to_csv('Segment_table', index = False)
Country_table.to_csv('Country_table', index = False)
Product_table.to_csv('Product_table', index = False)
Discount_Band_table.to_csv('Discount_Band_table', index = False)

In [169]:
# Melakukan NF dengan mereplace value menjadi (1,2,3,4,...)
df['Segment'] = df['Segment'].str.replace('Government','1').str.replace('Midmarket','2').str.replace('Channel Partners','3').str.replace('Enterprise','4').str.replace('Small Business','5')
df['Country'] = df['Country'].str.replace('Canada','1').str.replace('Germany','2').str.replace('France','3').str.replace('Mexico','4').str.replace('United States of America','5')
df['Product'] = df['Product'].str.replace(' Carretera ','1').str.replace(' Montana ','2').str.replace(' Paseo ','3').str.replace(' Velo ','4').str.replace(' VTT ','5').str.replace(' Amarilla ','6')
df['Discount Band'] = df['Discount Band'].str.replace(' None ','1').str.replace(' Low ','2').str.replace(' Medium ','3').str.replace(' High ','4')

In [170]:
# merubah data type dari object menjadi integer
df['Segment']  = df['Segment'].astype(int)
df['Country']  = df['Country'].astype(int)
df['Product'] = df['Product'].astype(int)
df['Discount Band'] = df['Discount Band'].astype(int)

In [171]:
# menampilkan tabel yang telah di cleaning, normalisasis 3NF
df

Segment  Country  Product  Discount Band  Units Sold  \
0          1        1        1              1      1618.5   
1          1        2        1              1      1321.0   
2          2        3        1              1      2178.0   
3          2        2        1              1       888.0   
4          2        4        1              1      2470.0   
..       ...      ...      ...            ...         ...   
695        5        3        6              4      2475.0   
696        5        4        6              4       546.0   
697        1        4        2              4      1368.0   
698        1        1        3              4       723.0   
699        3        5        5              4      1806.0   

     Manufacturing Price  Sale Price  Gross Sales  Discounts      Sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         COGS    Profit       Date                  Month Number  Month Name  \
0     16185.0  16185.00 2014-01-01 1970-01-01 00:00:00.000000001    January    
1     13210.0  13210.00 2014-01-01 1970-01-01 00:00:00.000000001    January    
2     21780.0  10890.00 2014-01-06 1970-01-01 00:00:00.000000006       June    
3      8880.0   4440.00 2014-01-06 1970-01-01 00:00:00.000000006       June    
4     24700.0  12350.00 2014-01-06 1970-01-01 00:00:00.000000006       June    
..        ...       ...        ...                           ...         ...   
695  618750.0  12375.00 2014-01-03 1970-01-01 00:00:00.000000003      March    
696  136500.0   2730.00 2014-01-10 1970-01-01 00:00:00.000000010    October    
697    6840.0   1299.60 2014-01-02 1970-01-01 00:00:00.000000002   February    
698    3615.0    686.85 2014-01-04 1970-01-01 00:00:00.000000004      April    
699    5418.0  13003.20 2014-01-05 1970-01-01 00:00:00.000000005        May    

     Year  
0    2014  
1    2014  
2    2014  
3    2014  
4    2014  
..    ...  
695  2014  
696  2014  
697  2014  
698  2014  
699  2014  

[700 rows x 16 columns]

In [172]:
# menampilkna tipe data setelah selesai di normalisasi dan diubah tipe datanya
df.dtypes

Segment                         int32
Country                         int32
Product                         int32
Discount Band                   int32
Units Sold                    float64
Manufacturing Price           float64
Sale Price                    float64
Gross Sales                   float64
Discounts                     float64
Sales                         float64
COGS                          float64
Profit                        float64
Date                   datetime64[ns]
Month Number           datetime64[ns]
Month Name                     object
Year                            int64
dtype: object

In [173]:
# mengsave tabel yang telah di cleaning, normalisasis 3NF
df.to_csv('tabel_lengkap', index = False)

# 4. Relational Database & SQL

dikerjakan di postgressql

# 5. Pengujian Database

## 1. Setelah tabel sudah berhasil dibuat, uji keberhasilan database dan tabel yang kamu buat dengan kasus berikut:

### a. Berikan tabel yang berisikan informasi total profit yang didapatkan di tiap jenis segmentasi. Jangan ambil data yang tidak diskon.

```sql
-- membuat tabel berisikan informasi total profit yang didapatkan di tiap jenis segmentasi dan tidak mengambil data yang tidak ada diskon
-- kemudian disave menggunakan query copy dan to
BEGIN;
copy (
SELECT Segment, SUM(profit) as total_profit
FROM tabel_lengkap
WHERE discounts > 0
GROUP BY 1
ORDER BY 1 asc
)
TO 'C:\tmp\pengujian_1' WITH CSV HEADER;
COMMIT;
```

### b. Berikan tabel yang berisikan informasi rangkuman statistik yang memuat nilai rata-rata, min, dan max (dijadikan dalam kolom yang berbeda) dari Sales masing-masing negara.

```sql
-- membuat tabel berisikan informasi rangkuman statistik yang memuat nilai rata-rata, min, dan max dari sales masing-masing negara
-- kemudian disave menggunakan copy dan to
BEGIN;
copy (
SELECT 
	Country,
	AVG(sales) as rata2_sales,
	MIN(sales) as min_sales,
	MAX(sales) as max_sales
FROM tabel_lengkap
GROUP BY 1
ORDER BY 1 asc
)
TO 'C:\tmp\pengujian_2' WITH CSV HEADER;
COMMIT;
```

## 2. Simpan kedua tabel dalam file yang berbeda dan load ke dalam notebook pengolahan data untuk ditampilkan. Tuliskan query SQL pada markdown!.

In [174]:
# memunculkan tabel yang telah disave menjadi csv dari postgressql
pengujian_1 = pd.read_csv('C:\\tmp\pengujian_1')
pengujian_1

segment  total_profit
0        1   10270950.18
1        2     563253.09
2        3    1150321.14
3        4     879096.88
4        5    3848268.50

didapatkan total profit dari setiap jenis segmentasi dan tidak mengambil data yang ada diskon, kemudian diurutkan berdasarkan segmentasi

`1` = `Government` dengan total profit `10270950.18`

`2` = `Midmarket` dengan total profit `563253.09`

`3` = `Channel Partners` dengan total profit `1150321.14`

`4` = `Enterprise` dengan total profit `879096.88`

`5` = `Small Business` dengan total profit `3848268.50`

In [175]:
# memunculkan tabel yang telah disave menjadi csv dari postgressql
pengujian_2 = pd.read_csv('C:\\tmp\pengujian_2')
pengujian_2

country    rata2_sales  min_sales  max_sales
0        1  177768.963500    1655.08  1038082.5
1        2  167895.291571    1685.60  1017338.0
2        3  173958.373500    1763.86   962500.0
3        4  149638.229357    2508.66   848172.5
4        5  178784.501286    1730.54  1159200.0

rangkuman statistik yang memuat nilai rata-rata, min, dan max dari sales masing-masing negara

`1` = `Canada` dengan rata2 sales `177768.963500`, sales terkecil `1655.08`, dan sales terbesar `1038082.5`

`2` = `Germany` dengan rata2 sales `167895.291571`, sales terkecil `1685.60	`, dan sales terbesar `1017338.0`

`3` = `France` dengan rata2 sales `173958.373500`, sales terkecil `1763.86`, dan sales terbesar `962500.0`

`4` = `Mexico` dengan rata2 sales `149638.229357`, sales terkecil `2508.66`, dan sales terbesar `848172.5`

`5` = `United States of America` dengan rata2 sales `178784.501286	`, sales terkecil `1730.54`, dan sales terbesar `1159200.0`


# Kesimpulan

garis besar yang telah saya lakukan adalah data exploration untuk mengetahui isi data, tipe data, berapa banyak data. kemudian data cleaning untuk merapikan data yang sudah ada seperti merapihkan nama kolom, merapihkan value kolom agar tipe datanya sesuai dengan yang di inginkan, menlengkapi value agar tidak ada missing value. kemudian data normalization, setelah itu data disave sebagai ccsv, lalu membuat databse di postgresSql dan data tadi dimasukan kedalam database tersebut. Dan yang terakhir adalah pengujian database apakah berjalan atau tidak atau apakah hasil yang diinginkan sudah sesuai.